homeassistant/components/mqtt/__init__.py

```diff
@@ -774,6 +801,20 @@ async def finish_dump(_):
         ),
     )
 
+    # setup platforms and discovery
+    hass.data[DATA_CONFIG_ENTRY_LOCK] = asyncio.Lock()
+    hass.data[CONFIG_ENTRY_IS_SETUP] = set()
+
+    yaml_platforms = await async_get_platform_components(hass, PLATFORMS_YAML_SETUP)
+    async with hass.data[DATA_CONFIG_ENTRY_LOCK]:
```

Comment:
We should probably just unconditionally load all MQTT platforms when MQTT is setup, then we can remove the code in discovery.py for loading MQTT platforms and also this lock.

That should be done in a separate PR though.

Reply:
Since we cannot await the setup, we do not known if the discovery setup is there sooner. Not all platforms support a config setup using configuration.yaml.